### ISSUE TITLE
[question] I was traumatized while trying to build the qt package. Is this a common problem and if so are there ideas on how building large dependency trees could be made more stable?

### ISSUE THREAD

**Issue by Knitschi** [conan-io/conan#7899]
# The Trauma
I am trying to integrate conan into a little private toy project that uses Qt. This project is build on Windows and Ubuntu. I build it in 5 different configurations *VS2019-shared-debug*, *VS2019-static-release*, *Clang-shared-debug*, *Clang-static-release* and *Gcc-shared-debug*. 

I tried all available versions of the Qt package starting with 5.12.6. That is 11 Versions. Guess what, none of them compiled for all of my configurations. Considering now that one Qt build takes quite some time, you can imagine that getting this data alone cost me a weekend. Also interesting about this is, that I countent at least 7 different reasons for the build to fail. Here are some of them:

* ```make[10]: *** [Makefile.xcb_qpa_lib:5240: .obj/qxcbkeyboard.o] Error 1``` Here I was not able to see what went wrong.
* ```ERROR: Conflict in pcre/8.41:``` At least a descriptive error message.
* ```ModuleNotFoundError: No module named '_lzma'```
* ```make[9]: *** [Makefile:789: .obj/qxkbcommon.o] Error 1```
* One was a low level python error that I could fix by updating conan.

My point here is, that the build was not broken by one problem, but by a large number of different problems. Often deep down in the dependency tree of the original project I wanted to build. These dependencies may be using build-systems that I am not familiar with. This makes deciphering the build-log in order to find the root cause of an error hard.

# Discussion
I know that this is not exactly a problem of conan or the conan qt package, but is rather caused by having large and very heterogenious dependency trees. I also have the feeling that the use of system libraries is also a source of constant trouble because they are not handled by conan or at least not consistently handled by conan. Then adding to that come the dependency version requirement conflicts. I think the only way to really solve those would be to allow multiple versions of one package to exist in the dependency tree. But at least cmake does not support that and I am not aware that any other build-system does.

 So in short ... It is a mess!

# Questions
* So are there any ideas on how building large heterogeneous dependency trees can be made more robust?
* Is this a problem that is discussed "up the food-chain" maybe in standard comitee circles?
* Are you guys, as the ones who have to suffer most from this problem, have any initiatives going that try to fix this?

Cheers
Knitschi

## References

The project that  I used to test the qt package and that contains the different conanfiles can be found at:
[ConanQtPackageTest](https://github.com/Knitschi/ConanQtPackageTest)

**Comment by memsharded** [comment]
Hi @Knitschi 

Qt is indeed a huge challenge. @ericLemanissier has been doing an amazing job, and there is a PR open to ConanCenter to add it: https://github.com/conan-io/conan-center-index/pull/1759.

I think it would be great to start from there, and not from the qt/xxx@bincrafters/stable packages. The PR to conan-center might be more modern and maintained, and validated against the latest ConanCenter. Please @ericLemanissier confirm.

As at the moment this is not an issue of Conan client per-se, we could move it to conan-center-index?

**Comment by ericLemanissier** [comment]
@Knitschi The problem with the situation is of a combinatorial nature. We simply cannot verify constantly that 12 different versions of qt work on 20 different compilers as (debug+release)*(shared+static), and I'm not even trying to count the os variations. Breakage can happen at any modifications of any package in the dependency tree, and this tree is quite important in the case of Qt. This is especially hard when the maintainer is doing it on his spare time. Conan-Center-Index recipes have the advantages of beeing tested for all versions at each changes, but it is not the model of bincrafters recipes.
You very boldly tried to build 12 different versions whereas most of bincrafters project truly maintain only the latest version. Qt itself does maintain only few branches. I recommend you to use only the version maintained upstream, which is the latest 5.12.x and 5.15.x. Please file bug reports with full build logs on https://github.com/bincrafters/community so that we can help you.
The best would be if you could add some public CI to https://github.com/Knitschi/ConanQtPackageTest

@memsharded thanks for your kind words, but unfortunately I cannot advise to use the PR recipe. It basically has no user because it is not published anywhere, and it is very severe limitations for cmake consumers, because of https://github.com/bincrafters/conan-center-index/blob/master/docs/faqs.md#why-are-cmake-findconfig-files-and-pkg-config-files-not-packaged. Interesting changes proposed by reviewers of the PR are usually re-implemented in https://github.com/bincrafters/conan-qt/ too.

**Comment by Knitschi** [comment]
Thank you for your replies. I can imagine that maintaining the Qt package on your spare time is a pain. I use it in my spare time and every day that I spend with hunting down qt build problems feels especially frustrating. 

The reason why I put this here and not in the bincrafter package repo is, that I have the feeling that this is not a problem that is specific to Qt, but maybe occurrs for all packages that have large dependency trees below them. Many of the errors that I encountered are not in Qt itself, but rather when compiling one of the dependencies.

I thouth that you as the package manager developers have a lot of experience with the problems that occurr and maybe you have the wisdom to provide library developers with a set of good practices on how they can make their builds more robust. I know that in the end you are not responsible to fix other teams builds.

The problem is that it such a slow process to get an issue fixed, because you first have to convince some other team that they have to fix their build for the configuration you want and than you also need to convince the package maintainer to create a new version of the package.

**Comment by ericLemanissier** [comment]
@Knitschi your build scripts seem to invoke conan instances in parallel. Maybe you hit a bug in conan concurrency. Do you have the same errors when you use the `--cpus 1` argument to https://github.com/Knitschi/CPFBuildscripts/blob/master/4_Make.py ?

**Comment by ericLemanissier** [comment]
@Knitsch: I just made a github actions which tries to build the latest patch versions of [the recipe you tried](https://github.com/Knitschi/ConanQtPackageTest/tree/master/Sources/QtTestConsumer), with all the profile you defined [in your project](https://github.com/Knitschi/CIBuildConfigurations). results are available here: https://github.com/ericLemanissier/ConanQtPackageTest/actions/runs/316417279. There is much less failure that what you encountered:
- two linux build failing with: `meson.build:805:2: ERROR: Problem encountered: Python (3.x) mako module >= 0.8.0 required to build mesa.` whichare solved by running `pip install mako`
- a Windows build failing with `freetype.lib(sfnt.obj) : error LNK2019: unresolved external symbol png_create_read_struct referenced in function Load_SBit_Png etc.`

EDIT: with mako installed, only one error remains: https://github.com/ericLemanissier/ConanQtPackageTest/actions/runs/316958173

Of the error you described, the only one I had is `checking for m4 that supports -I option... configure: error: could not find m4 that supports -I option` and it was solved by running `apt install m4`

all this would tend to confirm that a lot of your problems could be caused by a concurrency issue.

**Comment by Knitschi** [comment]
On the ubuntu-20.04 runner that is provided by Github almost all the package configurations seem to build fine. I guess I will install the same packages on my private build-client and hope that it fixes the problems.

`all this would tend to confirm that a lot of your problems could be caused by a concurrency issue.`

Does conan not allow running multiple instances at the same time?

**Comment by memsharded** [comment]
Closing this ticket as outdated, it has been a long time, Conan 1 has been replaced by Conan 2, and all the recipes in ConanCenter have been update and largely changed since then.

If there are any further question or issues, please create new tickets. Thanks for the feedback.
